keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+- If there are foreign finalizers on the Evacuation, it should only remove its own instigator
+  finalizer (see [Drain](#drain)).
+- If the evacuator does not support a cancellation and it has set
+  `.status.evacuationCancellationPolicy` to `Forbid`, deletion of the Evacuation object will not be
+  attempted.
+
+Consequences for pods:
+1. Pods whose evacuators have not yet initiated evacuation will continue to run unchanged.
+2. Pods whose evacuators have initiated evacuation and support cancellation
+   (`.status.evacuationCancellationPolicy=Allow`) should cancel the evacuation and keep the pods
+   available.
+3. Pods whose evacuators have initiated evacuation and do not support cancellation
+   (`.status.evacuationCancellationPolicy=Forbid`) should continue the evacuation and eventually
+   terminate the pods.
+
+#### Drain
+
+When a `Drain` stage is detected on the NodeMaintenance object, Evacuation objects are created for
+selected pods ([Pod Selection](#pod-selection)).
+
+```yaml
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
+  name: f5823a89-e03f-4752-b013-445643b8c7a0-muffin-orders-6b59d9cb88-ks7wb
+  namespace: blue-deployment
+spec:
+  podRef:
+    name: muffin-orders-6b59d9cb88-ks7wb
+    uid:  f5823a89-e03f-4752-b013-445643b8c7a0
+  progressDeadlineSeconds: 1800
+
+```
+
+This is resolved to the following Evacuation object according to the pod on admission:
+
+```yaml
+
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
+  labels:
+    app: muffin-orders
+  name: f5823a89-e03f-4752-b013-445643b8c7a0-muffin-orders-6b59d9cb88-ks7wb
+  namespace: blue-deployment
+spec:
+  podRef:
+    name: muffin-orders-6b59d9cb88-ks7wb
+    uid:  f5823a89-e03f-4752-b013-445643b8c7a0
+  progressDeadlineSeconds: 1800
+  evacuators:
+    - evacuatorClass: deployment.apps.k8s.io
+      priority: 10000
+      role: controller
+```
+
+The node maintenance controller requests the removal of a pod from a node by the presence of the
+Evacuation. Setting `progressDeadlineSeconds` to  1800 (30m) should give potential evacuators
+enough time to recover from a disruption and continue with the graceful evacuation. If the
+evacuators are unable to evacuate the pod, or if there are no evacuators, the evacuation controller
+will attempt to evict these pods, until they are deleted.
+
+The only job of the node maintenance controller is to make sure that the Evacuation object exist
+and has the `evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io` finalizer. 
+
+#### Pod Selection
+
+The pods for evacuation would first be selected by node (`.spec.nodeSelector`). NodeMaintenance
+should eventually remove all the pods from each node. To do this in a graceful manner, the
+controller will first ensure that lower priority pods are evacuated first for the same pod type.
```

Comment:
> If I understand correctly, this proposal assumes that the priority of Pods is also the (reverse) order in which Pods should be drained. I'm not sure if that is a correct assumption.

yes

> I wonder if it should be possible to define drain orders independent of Pod priorities.

We work with pod priorities in other places in kube and they are used by all workloads today. What would be the benefit of not using them?

Reply:
I think coupling the Pod priorities to drain order could be problematic:
1.) I think nobody today anticipates that the Pod priority is also the field that decides the drain order
  * I'm mostly not sure if it's safe the assume that the "prioritiy" of a Pod is also *always* the reverse order in which it should be drained
2.) What if folks don't have the permissions to adjust the priority of a Pod, but still want to be able to control drain order (this would easily possible with a pod & namespace selector because it doesn't require permissions to change the Pod itself)